tornadox/solvers/ek1.py

```diff
@@ -189,3 +189,232 @@ def extract_qoi_fn(t, state):
+            )
+            p = jnp.tile(p, ode_dimension)
+            p_inv = jnp.tile(p_inv, ode_dimension)
+
+            # Attempt step
+            rv_new, _, _, bw_model, error = ek1_projmatfree_d_nu.attempt_step(
+                f=f,
+                df=df,
+                m=m0,
+                c_sqrtm=c_sqrtm0,
+                p=p,
+                p_inv=p_inv,
+                a=a,
+                q_sqrtm=q_sqrtm,
+                num_derivatives=num_derivatives,
+            )
+            error = dt_clipped * error  # The solver step does not do this.
+
+            # Normalise the error
+            m_new, _ = rv_new
+            u_old = m0[0 :: (num_derivatives + 1)]
+            u_new = m_new[0 :: (num_derivatives + 1)]
+            u1_ref = jnp.abs(jnp.maximum(u_old, u_new))
+            error_rel = error / (atol + rtol * u1_ref)
+            error_norm = jnp.linalg.norm(error_rel) / jnp.sqrt(error.size)
+
+            # Propose a new time-step
+            scale_factor = stepsizes.scale_factor_pi_control(
+                error_norm=error_norm,
+                error_order=num_derivatives + 1,
+                safety=safety,
+                factor_min=factor_min,
+                factor_max=factor_max,
+                error_norm_previously_accepted=error_norm_previously_accepted,
+                power_integral_unscaled=power_integral_unscaled,
+                power_proportional_unscaled=power_proportional_unscaled,
+            )
+            dt_new = scale_factor * dt_clipped
+
+            stats = s_prev.stats
+            stats["f_evaluation_count"] += 1
+            stats["df_evaluation_count"] += 1
+            stats["steps_attempted_count"] += 1
+
+            s_new = EK1State(
+                u=rv_new,
+                backward_model=bw_model,
+                dt_proposed=dt_new,
+                error_norm=error_norm,
+                stats=stats,
+            )
+            return t_new, s_new
+
+        # While-loop until the error_norm is smaller than 1.
+        # For the final output, we discard the error_norm.
+        larger_than_1 = 1.1
+        init_val = EK1State(
+            u=state0.u,
+            backward_model=state0.backward_model,
+            dt_proposed=state0.dt_proposed,
+            error_norm=larger_than_1,
+            stats=state0.stats,
+        )
+        t, state = jax.lax.while_loop(
+            cond_fun=cond_fun, body_fun=body_fun, init_val=(t0, init_val)
+        )
+
+        # Condense both backward models into one.
+        # (Technically speaking, this implements 'fixed-point smoothing')
+        # todo: if we do this somewhere else, perform_step_fn() can be reused
+        #  for other solve_ivp()-style functions.
+        A, (b, B_sqrtm) = init_val.backward_model
+        C, (d, D_sqrtm) = state.backward_model
+        g = A @ C
+        xi = A @ d + b
+        Xi = sqrtutil.sum_of_sqrtm_factors(S1=A @ D_sqrtm, S2=B_sqrtm)
+        backward_model_fp = (g, (xi, Xi))
```

Comment:
I know we're not so consistent yet, but here I really find it helpful if we can have matrices capitals and vectors lower-case :sweat_smile: 
```suggestion
        G = A @ C
        xi = A @ d + b
        Xi = sqrtutil.sum_of_sqrtm_factors(S1=A @ D_sqrtm, S2=B_sqrtm)
        backward_model_fp = (G, (xi, Xi))
```

Reply:
I am not sure I am actually a fan of this capitalisation. But at this point here, I changed it in order to be consistent and to make it easier for you to read :)